draft-rfcs/const-evaluatable-checked.md

```diff
@@ -0,0 +1,12 @@
+# const evaluatable checked
+
+TODO: All of this :)
+
+## Requirements
+
+For this to work, const operations have to be both pure and deterministic, at least when they are used in the type system.
```

Comment:
"pure" is an observational notion, so the question here is -- pure for which kind of observer? (To give an example: Haskell is very impure if you go down to the level of lazily evaluated thunks and forcing.)

I think the best way to organize this is to have a "value tree" representation of CTFE results, and only that representation will be observed by the type system. Value trees contain no `AllocId`, which are the only source of impurity in CTFE. (And non-determinism does not arise in CTFE at all.) So I think on this level, CTFE *is* pure, even with allocations and all that. This means converting a CTFE result to a value tree can fail (this is a "dynamic check" that the CTFE result is indeed consumable in a pure way); you might want to have extra static checks to ensure that this dynamic check never fires -- basically, the return type must not contain raw pointers, fn pointers, floats, or anything else that we cannot convert to a value tree (trait objects?).

See [this issue](https://github.com/rust-lang/rust/issues/72396) and [this hackmd](https://hackmd.io/Qvrj_eOFTkCHZrhJ7f1ItA) for some more notes on value trees. Those are very terse, so if clarification is needed, please ask. :)

Reply:
yeah, "pure" in that the output of CTFE does not depend on any global state, reworded this. ^^

I am still not sure on how in depth I want to go here, so I think I am going to leave the document like this for now.

Thanks for the elaborate answers :heart: